fxa-oauth-server/lib/routes/token.js

```diff
@@ -139,221 +146,199 @@ module.exports = {
+      params.client_secret = Joi.attempt(creds[1], validators.clientSecret, err);
+    }
+
+    const client = await authenticateClient(params);
+    const requestedGrant = await validateGrantParameters(client, params);
+    return await generateTokens(requestedGrant);
   }
 };
 
-/**
- * Generate a PKCE code_challenge
- * See https://tools.ietf.org/html/rfc7636#section-4.6 for details
- */
-function pkceHash(input) {
-  return util.base64URLEncode(crypto.createHash('sha256').update(input).digest());
-}
 
-function validPublicClient(client) {
-  if (! client.publicClient) {
-    logger.debug('client.notPublicClient', { id: client.id });
-    throw AppError.notPublicClient(client.id);
+async function authenticateClient(params) {
+  const client = await getClientById(params.client_id);
+  // Public clients can't be authenticated in any useful way,
+  // and should never submit a client_secret.
+  if (client.publicClient) {
+    if (params.client_secret) {
+      throw new AppError.invalidRequestParameter('client_secret');
+    }
+    return client;
   }
+  // Check client_secret against both current and previous stored secrets,
+  // to allow for seamless rotation of the secret.
+  // N.B. non-constant-time comparison is safe here due to the hashing.
+  const submitted = encrypt.hash(buf(params.client_secret));
+  const stored = client.hashedSecret;
+  if (crypto.timingSafeEqual(submitted, stored)) {
+    return client;
+  }
+  const storedPrevious = client.hashedSecretPrevious;
+  if (storedPrevious) {
+    if (crypto.timingSafeEqual(submitted, storedPrevious)) {
+      logger.info('client.matchSecretPrevious', { client: client.id });
+      return client;
+    }
+  }
+  logger.info('client.mismatchSecret', { client: client.id });
+  logger.verbose('client.mismatchSecret.details', {
+    submitted: submitted,
+    db: stored,
+    dbPrevious: storedPrevious
+  });
+  throw AppError.incorrectSecret(client.id);
+}
 
-  return true;
+
+async function getClientById(clientId) {
+  const client = db.getClient(buf(clientId));
+  if (! client) {
+    logger.debug('client.notFound', { id: clientId });
+    throw AppError.unknownClient(clientId);
+  }
+  return client;
 }
 
-function getClientById(clientId) {
-  return db.getClient(buf(clientId)).then(function(client) {
-    if (! client) {
-      logger.debug('client.notFound', { id: clientId });
-      throw AppError.unknownClient(clientId);
-    }
 
-    return client;
-  });
+async function validateGrantParameters(client, params) {
+  let requestedGrant;
+  switch (params.grant_type) {
+  case GRANT_AUTHORIZATION_CODE:
+    requestedGrant = await validateAuthorizationCodeGrant(client, params);
+    break;
+  case GRANT_REFRESH_TOKEN:
+    requestedGrant = await validateRefreshTokenGrant(client, params);
+    break;
+  default:
+    // Joi validation means this should never happen.
+    logger.critical('joi.grant_type', params.grant_type);
```

Comment:
Same question here about Stackdriver logging and whether the 2nd param needs to be an object.